TRUNK-4702. OrderService should allow you to save overlapping DrugOrders

sintjuri opened the pull request:
if the caller acknowledges that they are aware of the overlap, otherwise
AmbiguousOrderException have to be thrown

https://issues.openmrs.org/browse/TRUNK-4702

wluyima commented on api/src/main/java/org/openmrs/api/OrderService.java:
```diff
@@ -85,7 +85,7 @@
 	 * @should set Order type of Test Order to test order if not set and concept not mapped
-	 * @should fail if an active drug order for the same drug formulation exists
+	 * @should throw AmbiguousOrderException if an active drug order for the same drug formulation exists
 	 * @should pass if an active order for the same concept exists in a different care setting
```
You also need a test to ensure that it passes for overlapping orders in case the parallel orders are specified

wluyima commented on api/src/main/java/org/openmrs/api/OrderService.java:
```diff
@@ -98,6 +98,20 @@
+	* @should pass if an known drug order for the same drug formulation specified
+	*/
+	@Authorized( { PrivilegeConstants.EDIT_ORDERS, PrivilegeConstants.ADD_ORDERS })
+	public Order saveOrder(Order order, OrderContext orderContext, Order[]  parallelOrders) throws APIException;
```
I think you should use the orderContext for now instead and not overload this method

wluyima replied:
I think the parallel orders should be uuids instead of order objects

sintjuri commented:
Hi Wyclif,

I changed array of Orders to array of UIDs. It is clear. I also removed overridden method 
saveOrder(Order order, OrderContext orderContext, Order[] parallelOrders)
despite of requirements in https://issues.openmrs.org/browse/TRUNK-4702. What about test is not clear. I have already test org.openmrs.api.OrderServiceTest.saveOrder_shouldPassIfAnKnownDrugOrderForTheSameDrugFormulationSpecified(). Should I write another one?

dkayiwa commented:
@wluyima what do you think of the above?

wluyima commented on api/src/main/java/org/openmrs/api/impl/OrderServiceImpl.java:
```diff
@@ -201,18 +202,22 @@ public synchronized Order saveOrder(Order order, OrderContext orderContext) thro
+ 			for (Order activeOrder : activeOrders) {
+ 				//Reject if there is an active drug order for the same orderable with overlapping schedule
+				if (!parallelOrders.contains(activeOrder.getUuid()) && areDrugOrdersOfSameOrderableAndOverlappingSchedule(order, activeOrder)) {
+					throw new AmbiguousOrderException("Order.cannot.have.more.than.one", (Object[]) null);
```
It is pointless to pass in a null value here when there is a constructor that takes no argument, i will remove it
